pkg/controllers/corecluster/corecluster_controller.go

```diff
@@ -81,38 +87,89 @@ func (r *CoreClusterController) SetupWithManager(mgr ctrl.Manager) error {
 }
 
 // Reconcile reconciles the core cluster object for the openshift-cluster-api namespace.
+//
+//nolint:funlen
 func (r *CoreClusterController) Reconcile(ctx context.Context, req reconcile.Request) (ctrl.Result, error) {
 	logger := log.FromContext(ctx).WithName(controllerName)
 	logger.Info("Reconciling core cluster")
 	defer logger.Info("Finished reconciling core cluster")
 
+	var err error
+
+	co, err := r.GetOrCreateClusterOperator(ctx)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to get or create cluster operator: %w", err)
+	}
+
+	// Populate conditions from the existing cluster operator status.
+	conditions := &co.Status.Conditions
+
+	defer func() {
+		// This function runs before every return, to ensure the controller conditions
+		// for that codepath are correctly synced to the cluster operator status.
+		if syncErr := r.SyncControllerConditions(ctx, co, controllerName, conditions,
+			[]string{CoreClusterControllerAvailableCondition, CoreClusterControllerDegradedCondition}); syncErr != nil {
+			if err != nil {
+				// The sync of the controller conditions failed but there is also a non nil reconcile error.
+				// Log the sync error and pass down the reconciler one.
+				logger.Error(syncErr, "failed to sync cluster operator status")
+			} else {
+				// If there is no reconcile error, but there is a sync error, then we want to propagate the
+				// sync error to be returned by the Reconcile function, so that it will cause an immediate requeue/retry.
+				err = syncErr
+			}
+		}
+	}()
+
 	ocpInfrastructureName, err := getOCPInfrastructureName(r.Infra)
 	if err != nil {
+		operatorstatus.SetCondition(conditions, CoreClusterControllerDegradedCondition, configv1.ConditionTrue,
+			operatorstatus.ReasonSyncFailed, fmt.Sprintf("controller is degraded: %s", err.Error()))
+
+		operatorstatus.SetCondition(conditions, CoreClusterControllerAvailableCondition, configv1.ConditionTrue,
+			operatorstatus.ReasonAsExpected, "controller is available")
+
```

Comment:
We can't say that the controller is available here, or in any of the error cases. What if we never fetched the infra, and so we haven't synced the core cluster at all?

We should only set available true at the end when there have been no errors, no?

Reply:
Again, this boils down to the actual meaning we want to give to Available=false vs Unknown.
Happy to stay unknown in these cases and only go available once we finally reach the bottom of the reconcile logic.

Reply:
Available for the CoreCluster controller means a CoreCluster is in the cluster and doesn't have a deletion timestamp
(which shouldn't happen anyway because admission validations block this), is there anything else it needs to be considered available? A ready condition?

So any path that errors, doesn't know if the core cluster is there right? We only know the core cluster is there and available when we have completed a successful reconcile, so IMO, it should only be available true right at the end of the reconcile, or, once we know that the corecluster exists (if that is earlier)